src/meshapi/views/forms.py

```diff
@@ -191,13 +190,44 @@ def join_form(request):
     )
 
 
+def get_next_free_nn() -> int:
+    defined_nns = set(
+        Install.objects.exclude(install_status=Install.InstallStatus.REQUEST_RECEIVED, network_number__isnull=True)
+        .order_by("install_number")
+        .values_list("install_number", flat=True)
+    ).union(set(Install.objects.values_list("network_number", flat=True)))
+
+    # Find the first valid NN that isn't in use
+    free_nn = next(i for i in range(NETWORK_NUMBER_MIN, NETWORK_NUMBER_MAX + 1) if i not in defined_nns)
+
+    # Sanity check to make sure we don't assign something crazy
+    if free_nn < NETWORK_NUMBER_MIN or free_nn > NETWORK_NUMBER_MAX:
+        raise ValueError(f"Invalid NN: {free_nn}")
+
+    return free_nn
+
+
+def acquire_lock_for_all_installs():
+    # This code may look useless but is extremely important. Here we use select_for_update()
+    # to acquire locks on every single row in the Install table. This is a bit hacky, but based
+    # on how our NN assignment logic works, we kind of need it. There's nothing else we can lock
+    # to prevent the same NN from being assigned to multiple installs when requested concurrently.
+    # We order by the table's primary key to prevent deadlocks due to out-of-order locking, and we
+    # call len() to make sure the queryset is traversed (which is what actually acquires the lock)
+    # This must be called in the context of a transaction.atomic block, which also automatically
+    # handles the release of the lock for us
+    lock_qs = Install.objects.select_for_update().all().order_by("install_number")
+    len(lock_qs)
```

Comment:
Very thorough explanation. I was gonna posit adding something like this to the join form, but that might not be necessary since the likelihood of collisions is so low. (Mostly thinking about what would happen if two members in the same building submitted the join form at the same time.) Thoughts?

Reply:
This isn't necessary on the join form, since the database handles the atomicity for us. Since the database automatically assigns the `install_number` and that is the table's primary key, it knows not to let two rows have the same value.

However, it is a good idea to confirm this with a test, so I'll add one

Reply:
Oh wait I see now, you're talking about simultaneous same-building signups. Hmmm

Reply:
Okay I add locks to the join form also